lenskit-svd/src/main/java/org/lenskit/mf/BPR/TrainingItemPair.java

```diff
@@ -0,0 +1,38 @@
+/*
+ * LensKit, an open source recommender systems toolkit.
+ * Copyright 2010-2016 LensKit Contributors.  See CONTRIBUTORS.md.
+ * Work on LensKit has been funded by the National Science Foundation under
+ * grants IIS 05-34939, 08-08692, 08-12148, and 10-17697.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as
+ * published by the Free Software Foundation; either version 2.1 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU General Public License for more
+ * details.
+ *
+ * You should have received a copy of the GNU General Public License along with
+ * this program; if not, write to the Free Software Foundation, Inc., 51
+ * Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+ */
+package org.lenskit.mf.BPR;
+
+/**
+ * class to represent a training pair in pairwise learning to rank items.
+ * By convention the pair represents the knowledge that g>l (according to user u)
+ */
+public class TrainingItemPair {
+    // REVIEW: if I make these ints (rows in the matrix as per the snapshot) the code would theoretically be faster.
```

Comment:
I like the generality of longs. Let's only make them `int`s if we measure a bottleneck.